### PR TITLE
feat: add a common API to integrate with hotswap tools

### PR DESCRIPTION
**Pull request by mcollovati**
## Description

Adds API to integrate with hotswap agents and to allow plugging class change reload plugins.

Part of #19261 
Part of #19262

## Type of change

- [ ] Bugfix
- [x] Feature

## Checklist

- [x] I have read the contribution guide: https://vaadin.com/docs/latest/guide/contributing/overview/
- [x] I have added a description following the guideline.
- [x] The issue is created in the corresponding repository and I have referenced it.
- [x] I have added tests to ensure my change is effective and works as intended.
- [x] New and existing tests are passing locally with my change.
- [ ] I have performed self-review and corrected misspellings.

#### Additional for `Feature` type of change

- [ ] Enhancement / new feature was discussed in a corresponding GitHub issue and Acceptance Criteria were created.

### CODE REVIEW

File: flow-server/src/main/java/com/vaadin/flow/router/internal/RouteRegistryHotswapper.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+                && hasNoComponentClasses(modifiedClasses)
+                && hasNoComponentClasses(removedClasses)) {
+            return false;
+        }
+        if (session.getAttribute(SessionRouteRegistry.class) != null) {
+            RouteUtil.updateRouteRegistry(
+                    SessionRouteRegistry.getSessionRegistry(session), Set.of(),
+                    modifiedClasses, removedClasses);
+        }
+        return false;
```
[the remark this answers is not in the document]
Comment (mcollovati) [reply flow-server/src/main/java/com/vaadin/flow/router/internal/RouteRegistryHotswapper.java:91]: Initially, I was thinking the same. But then I thought that even if the route registry gets updated, a page reload might not be required if the changed classes are currently not shown on the UI.
If they are, the Hotswapper will detect them and trigger a proper refresh/reload

**Review by caalador — COMMENTED** [review]
I did get class reloads to work quite ok and the route change did update the available route, but only as long as I have visited the route at least once.

The page doesn't update automatically for me at for UI changes and I need to manually refresh.

**Comment by mcollovati** [comment]
> I did get class reloads to work quite ok and the route change did update the available route, but only as long as I have visited the route at least once.

I noticed the same when the hotswap plugin is not initialized before the `RouteRegistryInitializer` is executed. Otherwise, changes to routes are applied even if not visited.
But probably there are other situations that prevent the plugin being notified about the changes.
Sometimes, I also noticed that even if I update and recompile a class, IDEA says: "0 classes to reload" (or something similar)

File: flow-server/src/main/java/com/vaadin/flow/router/internal/RouteRegistryHotswapper.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (mcollovati) [reply flow-server/src/main/java/com/vaadin/flow/router/internal/RouteRegistryHotswapper.java:91]: Good point, thanks! With your suggestion, the code will be easier to understand and maintain.

[the remark this answers is not in the document]
Comment (mcollovati) [reply flow-server/src/main/java/com/vaadin/flow/router/internal/RouteRegistryHotswapper.java:87]: done

**Review by caalador — APPROVED** [review]
Work much better than the existing one.
Could have another pair of eyes to look through the additions.

**Comment by mshabarov** [comment]
Tested in conjunction with [Vaadin HotSwapAgent plugin enhancement](https://github.com/mcollovati/HotswapAgent/tree/refactor/vaadin-plugin-for-24.5-api) and the live reload worked for me very well, without any [class loader exceptions](https://github.com/vaadin/flow/issues/19247) and with stable UI updates. 

Let me now review the code.

File: flow-server/src/main/java/com/vaadin/flow/router/internal/RouteRegistryHotswapper.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+ * <p>
+ * </p>
+ * This class is meant to be used in combination wit Flow
+ * {@link com.vaadin.flow.hotswap.Hotswapper} to immediately update routes
+ * registries when classes have been added or modified.
+ * <p>
+ * </p>
+ * For internal use only. May be renamed or removed in a future release.
+ *
+ * @since 24.4
```
Comment (mshabarov) [inline flow-server/src/main/java/com/vaadin/flow/router/internal/RouteRegistryHotswapper.java:41]:
```suggestion
 * @since 24.5
```

**Review by mshabarov — APPROVED** [review]

**Review by mshabarov — APPROVED** [review]

**Pull request merged**